File: app/services/balance_calculator.py
"""
Service for calculating balances within a group.
"""
from uuid import UUID
from app.models import Group, Balance, Expense, Member


class BalanceCalculator:
    """
    Calculates individual and group balances based on expenses.

    The balance calculation considers:
    - Total amount paid by each member
    - Total amount owed by each member (their share of expenses they participated in)
    - Net balance (paid - owed)
    """

    def calculate_balances(self, group: Group) -> list[Balance]:
        """
        Calculate the balance for each member in the group.

        Args:
            group: The group to calculate balances for

        Returns:
            List of Balance objects for each member
        """
        if not group.members:
            return []

        member_map = {member.id: member for member in group.members}

        # Initialize tracking dictionaries
        total_paid: dict[UUID, float] = {member.id: 0.0 for member in group.members}
        total_owed: dict[UUID, float] = {member.id: 0.0 for member in group.members}

        # Process each expense
        for expense in group.expenses:
            self._process_expense(expense, total_paid, total_owed)

        # Create balance objects
        balances = []
        for member in group.members:
            paid = round(total_paid[member.id], 2)
            owed = round(total_owed[member.id], 2)
            net = round(paid - owed, 2)

            balances.append(Balance(
                member_id=member.id,
                member_name=member.name,
                total_paid=paid,
                total_owed=owed,
                net_balance=net
            ))

        return balances

    def _process_expense(
        self,
        expense: Expense,
        total_paid: dict[UUID, float],
        total_owed: dict[UUID, float]
    ) -> None:
        """
        Process a single expense and update the running totals.

        Args:
            expense: The expense to process
            total_paid: Dictionary tracking total paid by each member
            total_owed: Dictionary tracking total owed by each member
        """
        # Credit the payer
        if expense.payer_id in total_paid:
            total_paid[expense.payer_id] += expense.amount

        # Calculate each participant's share
        num_participants = len(expense.participant_ids)
        if num_participants == 0:
            return

        share_per_person = expense.amount / num_participants

        # Debit each participant
        for participant_id in expense.participant_ids:
            if participant_id in total_owed:
                total_owed[participant_id] += share_per_person

    def get_net_balances(self, group: Group) -> dict[UUID, float]:
        """
        Get simplified net balances for debt simplification.

        Args:
            group: The group to get net balances for

        Returns:
            Dictionary mapping member_id to net balance
        """
        balances = self.calculate_balances(group)
        return {balance.member_id: balance.net_balance for balance in balances}

File: src/main/java/com/expensesplitter/service/BalanceCalculator.java
package com.expensesplitter.service;

import com.expensesplitter.dto.BalanceResponse;
import com.expensesplitter.model.Expense;
import com.expensesplitter.model.Group;
import com.expensesplitter.model.Member;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.*;

/**
 * Service for calculating balances within a group.
 *
 * Calculates how much each member has paid, how much they owe,
 * and their net balance (positive = others owe them, negative = they owe others).
 */
@Service
public class BalanceCalculator {

    private static final int SCALE = 2;
    private static final RoundingMode ROUNDING_MODE = RoundingMode.HALF_UP;

    /**
     * Calculates the balance for each member in the group.
     *
     * @param group the group to calculate balances for
     * @return list of balance responses for each member
     */
    public List<BalanceResponse> calculateBalances(Group group) {
        if (group.getMembers().isEmpty()) {
            return Collections.emptyList();
        }

        Map<UUID, Member> memberMap = new HashMap<>();
        Map<UUID, BigDecimal> totalPaid = new HashMap<>();
        Map<UUID, BigDecimal> totalOwed = new HashMap<>();

        // Initialize maps
        for (Member member : group.getMembers()) {
            memberMap.put(member.getId(), member);
            totalPaid.put(member.getId(), BigDecimal.ZERO);
            totalOwed.put(member.getId(), BigDecimal.ZERO);
        }

        // Process each expense
        for (Expense expense : group.getExpenses()) {
            processExpense(expense, totalPaid, totalOwed);
        }

        // Build balance responses
        List<BalanceResponse> balances = new ArrayList<>();
        for (Member member : group.getMembers()) {
            UUID memberId = member.getId();
            BigDecimal paid = totalPaid.get(memberId).setScale(SCALE, ROUNDING_MODE);
            BigDecimal owed = totalOwed.get(memberId).setScale(SCALE, ROUNDING_MODE);
            BigDecimal net = paid.subtract(owed).setScale(SCALE, ROUNDING_MODE);

            balances.add(new BalanceResponse(
                    memberId,
                    member.getName(),
                    paid,
                    owed,
                    net
            ));
        }

        return balances;
    }

    /**
     * Gets net balances as a map for debt simplification.
     *
     * @param group the group to calculate net balances for
     * @return map of member ID to net balance
     */
    public Map<UUID, BigDecimal> getNetBalances(Group group) {
        Map<UUID, BigDecimal> netBalances = new HashMap<>();

        for (BalanceResponse balance : calculateBalances(group)) {
            netBalances.put(balance.getMemberId(), balance.getNetBalance());
        }

        return netBalances;
    }

    /**
     * Processes a single expense and updates the running totals.
     *
     * @param expense   the expense to process
     * @param totalPaid map tracking total paid by each member
     * @param totalOwed map tracking total owed by each member
     */
    private void processExpense(Expense expense,
                                Map<UUID, BigDecimal> totalPaid,
                                Map<UUID, BigDecimal> totalOwed) {
        // Credit the payer
        UUID payerId = expense.getPayerId();
        if (totalPaid.containsKey(payerId)) {
            totalPaid.put(payerId, totalPaid.get(payerId).add(expense.getAmount()));
        }

        // Calculate share per participant
        List<UUID> participantIds = expense.getParticipantIds();
        if (participantIds.isEmpty()) {
            return;
        }

        BigDecimal sharePerPerson = expense.getAmount()
                .divide(BigDecimal.valueOf(participantIds.size()), SCALE + 4, ROUNDING_MODE);

        // Debit each participant
        for (UUID participantId : participantIds) {
            if (totalOwed.containsKey(participantId)) {
                totalOwed.put(participantId, totalOwed.get(participantId).add(sharePerPerson));
            }
        }
    }
}
